packages/controller/lib/setup/setupUpload.js

```diff
@@ -743,64 +710,67 @@ function Upload(options) {
             callback = logger;
             logger = null;
         }
+        return this.upgradeAdapterObjectsAsync(name, iopack, logger)
+            .then(() => callback && callback(name))
+            .catch(err => {
+                logger.error('Cannot upgrade Adapter Objects: ' + err);
+                callback && callback();
+            });
+    };
 
+    this.upgradeAdapterObjectsAsync = async (name, ioPack, logger) => {
         logger = logger || console;
 
         const adapterDir = tools.getAdapterDir(name);
-        let iopackFile;
+        let ioPackFile;
         try {
-            iopackFile = fs.readJSONSync(adapterDir + '/io-package.json');
+            ioPackFile = fs.readJSONSync(adapterDir + '/io-package.json');
         } catch {
             if (adapterDir) {
                 logger.error(`Cannot find io-package.json in ${adapterDir}`);
             } else {
                 logger.error(`Cannot find io-package.json for "${name}"`);
             }
-            iopackFile = null;
+            ioPackFile = null;
         }
-        iopack = iopack || iopackFile;
+        ioPack = ioPack || ioPackFile;
 
-        if (!iopack || iopack.common && iopack.common.controller) {
-            // if no iopack found or someone passed the controller
-            callback(name);
-        } else {
+        if (ioPack) {
             // Always update installed From from File on disk if exists and set
-            if (iopackFile && iopackFile.common && iopackFile.common.installedFrom) {
-                iopack.common = iopack.common || {};
-                iopack.common.installedFrom = iopackFile.common.installedFrom;
+            if (ioPackFile && ioPackFile.common && ioPackFile.common.installedFrom) {
+                ioPack.common = ioPack.common || {};
+                ioPack.common.installedFrom = ioPackFile.common.installedFrom;
+            }
+            // Not existing? Why ever ... we recreate
+            const obj = (await objects.getObject('system.adapter.' + name)) || {};
```

Comment:
I am not a fan of that, at the latest if we need to wrap try catch around it we would need to split, would just make two lines out of it